app/src/main/java/com/google/android/sambadocumentsprovider/mount/BrowsingAutocompleteAdapter.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright 2017 Google Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.google.android.sambadocumentsprovider.mount;
+
+import android.widget.Filter;
+import android.widget.Filterable;
+
+import com.google.android.sambadocumentsprovider.R;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+public class BrowsingAutocompleteAdapter extends SectionedAdapter implements Filterable {
+  private final List<ServerInfo> mServers = new ArrayList<>();
+
+  private List<ServerInfo> mFilteredServers = new ArrayList<>();
+
+  public void addServer(String name, List<String> children) {
```

Comment:
Actually instead of adding each server and its children to adapter, you should wire this class to a pure model class.

This adapter should register a listener to that model class.

I'll let you change that in the following patch.